src/dmd/timetrace_sema.d

```diff
@@ -0,0 +1,239 @@
+/**
+Visitors for adding timetracing code to DMD's semantic analysis visitors.
+
+This file is originally from LDC (the LLVM D compiler).
+
+Copyright: Copyright (C) 1999-2022 by The D Language Foundation, All Rights Reserved
+Authors:   Johan Engelen, Max Haughton
+License:   $(LINK2 https://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+Source:    $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/timetrace_sema.d, _timetrace.d)
+Documentation: https://dlang.org/phobos/dmd_timetrace.html
+Coverage:    https://codecov.io/gh/dlang/dmd/src/master/src/dmd/timetrace.d
+ */
+module dmd.timetrace_sema;
+
+import dmd.aggregate;
+import dmd.aliasthis;
+import dmd.arraytypes;
+import dmd.astcodegen;
+import dmd.attrib;
+import dmd.blockexit;
+import dmd.clone;
+import dmd.compiler;
+import dmd.dcast;
+import dmd.dclass;
+import dmd.declaration;
+import dmd.denum;
+import dmd.dimport;
+import dmd.dinterpret;
+import dmd.dmangle;
+import dmd.dmodule;
+import dmd.dscope;
+import dmd.dstruct;
+import dmd.dsymbol;
+import dmd.dtemplate;
+import dmd.dversion;
+import dmd.errors;
+import dmd.escape;
+import dmd.expression;
+import dmd.expressionsem;
+import dmd.func;
+import dmd.globals;
+import dmd.id;
+import dmd.identifier;
+import dmd.init;
+import dmd.initsem;
+import dmd.hdrgen;
+import dmd.mtype;
+import dmd.nogc;
+import dmd.nspace;
+import dmd.objc;
+import dmd.opover;
+import dmd.parse;
+import dmd.root.filename;
+import dmd.common.outbuffer;
+import dmd.root.rmem;
+import dmd.root.rootobject;
+import dmd.semantic2;
+import dmd.semantic3;
+import dmd.sideeffect;
+import dmd.statementsem;
+import dmd.staticassert;
+import dmd.tokens;
+import dmd.root.utf;
+import dmd.utils;
+import dmd.statement;
+import dmd.target;
+import dmd.templateparamsem;
+import dmd.typesem;
+import dmd.visitor;
+import dmd.common.timetrace;
+
+template pretext(string i)
+{
+    static if (i == "DsymbolSemanticVisitor")
+        private enum pretext = "Sema1: ";
+    else static if (i == "Semantic2Visitor")
+        private enum pretext = "Sema2: ";
+    else static if (i == "Semantic3Visitor")
+        private enum pretext = "Sema3: ";
+    else
+        static assert (false, "did not recognize SemaVisitor ==" ~ SemaVisitor);
+}
+/// Time tracing visitor for semantic analysis. Only valid to instantiate and use this visitor if time tracing is enabled.
+extern(C++) final class SemanticTimeTraceVisitor(SemaVisitor) : Visitor
```

Comment:
Repeated: This design was chosen to be able to non-invasively add tracing of the frontend. Now that tracing becomes part of the frontend, a different design choice could be made (e.g. add tracing code directly in SemanticXVisitor), perhaps later.
I consider this a little bit as a hack.